### PR TITLE
bootstrap: Don't apply -Ztls-model=initial-exec to deps of proc-macros

### LINKED ISSUE #134863: Dependency of rustc_fluent_macros getting compiled with -Ztls-model=initial-exec
Proc macros must not be compiled with `-Ztls-model=initial-exec` as `-Ztls-model=initial-exec` is incompatible with dlopen. Glibc is somewhat lenient with dlopening dylibs compiled with `-Ztls-model=initial-exec` for as long as there is room, but other libc implementations may not provide enough room or provide room at all. For example on FreeBSD trying to use rustc_driver results in the following error:

https://cirrus-ci.com/task/4557776715776000

```
error: /.rustup/toolchains/nightly-2024-12-28-x86_64-unknown-freebsd/lib/rustlib/x86_64-unknown-freebsd/lib/librustc_fluent_macro-60e8ff74c2f2aaa5.so: No space available for static Thread Local Storage
  --> src/lib.rs:16:1
   |
16 | extern crate rustc_middle;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^
```

To fix this issue, the set of exclusions at https://github.com/rust-lang/rust/blob/ceb0441e82259c6fca91aa5aea391e9d0f3ec272/src/bootstrap/src/bin/rustc.rs#L166-L173 needs to be updated to include all direct and indirect dependencies of the rustc_fluent_macros crate.

### PR DESCRIPTION
**Pull request by clubby789**
Fixes #134863

1. Checks if a crate name is in a static list before applying the flag
2. Adds a tidy check that gathers transitive deps of proc macros and ensures the list is up to date

cc @bjorn3 - the issue specifies `rustc_fluent_macro` but I assume this applies to all proc macro crates.

### CODE REVIEW

File: src/bootstrap/src/utils/proc_macro_deps.rs [hunk 1]
```diff
@@ -0,0 +1,153 @@
+/// Holds all direct and indirect dependencies of proc-macro crates in tree. See #134863
+pub static CRATES: &[&str] = &[
+    // tidy-alphabetical-start
+    "annotate-snippets",
+    "anstyle",
+    "autocfg",
```
Comment (onur-ozkan) [inline src/bootstrap/src/utils/proc_macro_deps.rs:6]: If we are going to add a tidy check for a list, I think it's essential to have the blessing functionality on that.

File: src/bootstrap/src/utils/proc_macro_deps.rs [hunk 2]
```diff
@@ -0,0 +1,153 @@
+/// Holds all direct and indirect dependencies of proc-macro crates in tree. See #134863
```
Comment (onur-ozkan) [inline src/bootstrap/src/utils/proc_macro_deps.rs:1]:
```suggestion
/// Holds all direct and indirect dependencies of proc-macro crates in tree.
/// See https://github.com/rust-lang/rust/issues/134863
```

File: src/tools/tidy/src/deps.rs [hunk 3]
```diff
@@ -600,6 +606,69 @@ pub fn check(root: &Path, cargo: &Path, bad: &mut bool) {
...
+/// Ensure the list of proc-macro crate transitive dependencies is up to date
+fn check_proc_macro_dep_list(root: &Path, cargo: &Path, bless: bool, bad: &mut bool) {
+    let mut cmd = cargo_metadata::MetadataCommand::new();
+    cmd.cargo_path(cargo)
+        .manifest_path(root.join("Cargo.toml"))
+        .features(cargo_metadata::CargoOpt::AllFeatures)
+        .other_options(vec!["--locked".to_owned()]);
+    let metadata = t!(cmd.exec());
+    let is_proc_macro_pkg =
+        |pkg: &&Package| pkg.dependencies.iter().any(|dep| dep.name.as_str() == "proc-macro2");
```
Comment (bjorn3) [inline src/tools/tidy/src/deps.rs:618]: Maybe check `pkg.targets.iter().any(|target| target.kind.iter().any(|kind| kind == TargetKind::ProcMacro))`? Not ever proc-macro necessarily depends on the proc-macro2 crate and not every crate that depends on the proc-macro2 crate is necessarily a proc-macro.

File: src/bootstrap/src/utils/proc_macro_deps.rs [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    "rand_core",
+    "rayon",
+    "rayon-core",
+    "redox_syscall",
+    "rinja_derive",
+    "rinja_parser",
+    "rustc-hash",
+    "rustc-rayon",
+    "rustc-rayon-core",
+    "rustc_fluent_macro",
```
Comment (bjorn3) [inline src/bootstrap/src/utils/proc_macro_deps.rs:90]: This list seems to include the proc-macros themself too, which is not really necessary as bin/rustc.rs already checks for `--crate-type proc-macro`. If it isn't too much effort, removing them from the list may slightly reduce the amount of times this list needs to be updated.

File: src/tools/tidy/src/deps.rs (hunk 3 above)
Comment (clubby789) [reply src/tools/tidy/src/deps.rs]: Ah this wasn't possible in the version of cargo_metadata we were using 😄 Updated to get this; good spot

**Review by onur-ozkan — APPROVED** [review]
Thanks!

**Comment by matthiaskrgr** [comment]
@bors r-
seems to have failed in https://github.com/rust-lang/rust/pull/134896#issuecomment-2564827577

File: src/tools/tidy/src/deps.rs [hunk 5]
```diff
@@ -600,6 +606,71 @@ pub fn check(root: &Path, cargo: &Path, bad: &mut bool) {
...
+        let mut proc_macro_deps: Vec<_> = proc_macro_deps_iter.collect();
+        proc_macro_deps.sort();
+        proc_macro_deps.dedup();
+        let mut file = File::create(root.join("src/bootstrap/src/utils/proc_macro_deps.rs"))
+            .expect("`proc_macro_deps` should exist");
+        writeln!(
+            &mut file,
+            "/// Do not update manually - use `./x.py test tidy --bless`
+/// Holds all direct and indirect dependencies of proc-macro crates in tree.
+/// See https://github.com/rust-lang/rust/issues/134863
```
Comment (lqd) [inline src/tools/tidy/src/deps.rs:637]: Probably

```suggestion
/// See <https://github.com/rust-lang/rust/issues/134863>
```

here and in the blessed output

Comment (clubby789) [reply src/tools/tidy/src/deps.rs]: Fixed; `x doc tidy` passes now

**Pull request merged**